[10.0][MIG]account_asset_management suite refactoring + 10.0 port

luc-demeyer opened the pull request:
Forward port of https://github.com/OCA/account-financial-tools/pull/501 helped thereby with the work done by @adrienpeiffer in https://github.com/OCA/account-financial-tools/pull/441.

Also new functionality (account_asset_management_import & account_asset_management_method_number_end).

Refactored account_asset_management_xls modules (no more cr.execute).
The excel reporting module depends on report_xlsx_helper (which is the functional equivalent of the 8.0 report_xls module).
Cf. https://github.com/OCA/reporting-engine/pull/204. I would be grateful if someone could merge that work so that this PR (as well as https://github.com/OCA/account-financial-reporting/pull/449) can pass travis.

luc-demeyer committed:
[10.0][MIG]account_asset_management suite refactoring + 10.0 port

luc-demeyer committed:
bugfix

luc-demeyer commented:
@benwillig 
Unit tests of module account_lock_date_update fails.
When I try to reproduce locally than this same unit test works fine.
Can you have a look so that I can end with a green PR ?

@adrienpeiffer 
Can you review this PR to ensure I have included all the work performed by you on PR #441 ?
Can we close PR #441 and focus on this one ?

sbidoul commented:
Hi @luc-demeyer thanks for this huge work!

We'll try to set aside some time to test and review. I agree to close #441 and focus on this one.

I've two questions:
- do you know if it can it be installed on a db which has #441 (same data model)?
- would you agree to make a separate PR for the core module to facilitate piece-wise review?

luc-demeyer commented:
@sbidoul 
>>> do you know if it can it be installed on a db which has #441 (same data model)?
A module upgrade should be sufficient.

>>> would you agree to make a separate PR for the core module to facilitate piece-wise review?
I prefer to keep this one PR. Having a asset management module without reporting doesn't make a lot of sense and the two other modules are very small hence easy to review. 
All modules come with a rather complete unit tests coverage and the code is very close to what we are running at multiple customer sites for over a year (on 8.0). I therefor feel rather comfortable on the quality of this work (a bit of help to diagnose the travis error on the account_lock_date_update module is highly appreciated, it looks unrelated to this PR but I don't see what's causing the issue and didn't manage to reproduce it).

luc-demeyer committed:
bugfix in raise

sanrav commented:
Hello,
There is an issue in the depreciation board when I click on "Compute" in the asset form.
I imported my asset (see Excel file enclosed) in the menu Accounting/Assets/Assets => Import. As it's an existing asset, there are 2 lines in the depreciation board : 1 for the depreciation base and 1 for the cumulative amortization. For both lines, the flag "Initial balance entry" ("Init") is set automatically with the import. Until there everything is fine. But when I click on "Compute", ODOO generates the depreciation lines (what is good) but set them as "Init". This means that when I confirm my asset, the status will be automatically "Close" and no amortization entries will be booked. 
I tried also to import through the menu "Fixed assets import" in the menu Accounting/Configuration/Assets but I have the same issue with the "Init" and moreover ODOO didn't create the line with the cumulative amortization.

[Test import assets 17 09 18.xlsx](https://github.com/OCA/account-financial-tools/files/2388317/Test.import.assets.17.09.18.xlsx)

sbidoul commented:
Hi @luc-demeyer, do you have an idea about the cause of the problem described by @sanrav above? That kind of import file did work well in v8 and with #441 to load assets that are already partly depreciaated.

luc-demeyer committed:
fix init flag when computing depreciation board

luc-demeyer commented:
@sanrav @sbidoul 
1) the init flag should only be set on entries in closed fiscal years (company.fiscalyear_lock_date field). I fixed the code so it should work now.
2) it's much easier to import assets via the import wizard in the account_asset_management_import module.

sbidoul commented on account_asset_management/models/account_asset.py:
```diff
@@ -895,7 +895,7 @@ def _compute_depreciation_table(self):
         fy = company.find_daterange_fy(asset_date_start)
         fiscalyear_lock_date = company.fiscalyear_lock_date
-        if fiscalyear_lock_date > self.date_start:
+        if fiscalyear_lock_date >= self.date_start:
```
@luc-demeyer Thanks for the fix.

BTW, why don't you use [compute_fiscalyear_dates](https://github.com/OCA/account-financial-tools/blob/29d69c94b71c2d7f3e277dfca68b7f2ed71ce2e4/account_fiscal_year/models/res_company.py#L27) to get the fiscal year end date? That methods supports the standard Odoo method (12 month FY) and is improved when account_fiscal_year is installed.

luc-demeyer replied:
@sbidoul 
The 'compute_fiscalyear_dates' falls back on the standard odoo method when no date_range of type fiscalyear is defined. That logic is not correct since it relies on the single fiscalyear_last_month/day fields whereas a company may have multiple changes in fiscal year duration.
We should as a consequence need to fix the logic in the account_fiscalyear OCA module. Maybe a good idea for the OCA code sprint next week (akim will work with me on the 11.0 port of the account_asset_management modules next week hence we could add this into the 11.0 port, @grindtildeath is this ok for you ?).

grindtildeath replied:
@luc-demeyer No issue to work on that from my side. :wink: 

BTW, was the refactoring here too big to keep commit history from v9.0 ?

sanrav commented:
Hello, the depreciation board works fine now. Thanks ! 
I uploaded my assets from the list view of the assets (standard import menu) because the menu "Fixed assets import" under Accounting/Configuration/Assets doesn't take into account the accumulated depreciated amount. That menu could be very interesting because we can choose to compute the depreciation board at the time of the import. Very useful if you have to import a lot of assets !  
A last small remark... When the journal entry is booked, it takes the name of the asset as journal entry number instead of a sequence. It's not a blocking point (we worked with that kind of entry number in my previous jobs) but could be changed in the future. The name of the asset is already in the field "Reference" of the journal entry.

luc-demeyer commented:
@sanrav 

1) "Fixed assets import" doesn't take into account the accumulated depreciated amount.

Indeed. In certain cases this is not an issue since the module should compute the same depreciated amount as the accounting package you are migrating from. But in some cases we see rounding deviations. In projects with massive amounts of assets I fixed aligned these deviations a script, in smaller project I instruct the customer to open the assets one by one and adjust manually the last 'init' entry amount. 
Adding this feature to the 'Fixed asset import' module would be useful. Feel free to contribute.

2) accounting entry name

The standard behaviour is on purpose since you can immediately see the link between the asset and the depreciation table entry in the accounting entry. The number and reference fields remain the same when removing and regenerating entries to correct encoding errors.
There is no legal obligation on consecutive numbering for the assets journal.
Hence you just need to spend a couple of minutes to explain this to your customer and afterwards they tend to like it more than their previous approach.
In case they still don't like it, you can use the 'def _setup_move_data(self, depreciation_date, period)' method to call the sequence defined on the journal (and at the end of the fiscalyear you may have to run the 'account_renumber' oca module to fix the numbering, but in imho it's a bad idea.
If you unable to convince your customer, than it's probably someone who is not ready to change old habits (and hence a risk on the ERP project because ERP projects tend to fail when customers are unwilling to change).

sanrav commented:
One last thing... 
![asset hierarchy](https://user-images.githubusercontent.com/39826575/46014398-616b8600-c0cf-11e8-8d12-1a1f93d19541.JPG)
In the list view of the assets and in the asset hierarchy, the total on the "view" assets is not calculated. It should show the total of the children assets. See screenshot enclosed.

luc-demeyer commented:
@sanrav 
The totals on the view assets was a stupid mistake from my side when I made the first version of the account_asset_management module in version 6.
Such a total is meaningless since assets get depreciated, removed, revaluated, ...
It only gives a confusing and meaningless figure and the calculation to compute this meaningless figure was a performance killer.
These totals are only meaningfull when making a fiscal year based report and that is exactly what the 'account_asset_management_xls' module does.

sanrav commented:
@luc-demeyer ok understood but you can maybe hide the zeros on the "view" assets because showing it like that gives the impression that something is not calculated properly.

luc-demeyer committed:
hide fields for type view assets

luc-demeyer commented:
I have now adapted the tree and form views to hide all non-necessary fields from type view assets.
I also removed the 'state' field since handling states on type view accounts doesn't make any sense.
Another improvement is that I do not create an asset line of type 'Depreciation Base' any more for assets with a zero value depreciation base.

sanrav commented:
The list view of the assets shows now twice the field "Residual value". One of them should be replaced by the field "Depreciated value".
![assets list view](https://user-images.githubusercontent.com/39826575/46132216-c644ee80-c23d-11e8-8aae-5cfa5ba6af2b.JPG)
In the assets hierarchy, ok there is no value on the "view" asset but there is no value anymore on the "normal" asset. Can we maybe just hide this Asset hierarchy menu ?! Because we have the same information in the list view of assets and there we can use "group by" to show the same result as assets hierarchy. 
![asset hierarchy v2](https://user-images.githubusercontent.com/39826575/46132526-8e8a7680-c23e-11e8-8390-0da94a525f70.JPG)

luc-demeyer committed:
fix viexs

luc-demeyer commented:
I fixed the tree view .
Concerning the hierarchy view: there is a little bit of value there as an 'aid' to visually the hierarchy (which comes back into the reporting). I simplified this view so that only this visualization is left and I also have limited the access to this view to the Financial Manager profile.

sanrav commented:
Ok great thanks. Everything looks good now.

luc-demeyer commented:
I think this one is ready for merge.
@sanrav 
during the OCA code sprint next week I'll work with C2C on the port of this work to v11.
Are you available and prepared to also review this work ?

grindtildeath commented on account_asset_management/__manifest__.py:
```diff
@@ -1,52 +1,35 @@
     'website': 'http://www.noviat.com',
     'category': 'Accounting & Finance',
     'sequence': 32,
-    'demo': ['account_asset_demo.xml'],
```
The file (https://github.com/OCA/account-financial-tools/blob/c56a8f963b03b8ba7318b38abfdbbea5e89db7fa/account_asset_management/demo/account_asset_demo.xml) is still there, I'd say remove the file if we don't need it, or keep the key in the manifest if we need it.

grindtildeath commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1075 @@
+    date_start = fields.Date(
+        string='Asset Start Date', readonly=True,
+        states={'draft': [('readonly', False)]},
+        default=fields.Datetime.now,
```
missing `()`

grindtildeath commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1075 @@
+        help="The number of years needed to depreciate your asset")
+    method_period = fields.Selection(
+        selection=lambda self: self.env['account.asset.profile'].
+        _selection_method_period(), string='Period Length',
```
would be great not to have the function on a new line :

        selection=lambda self: 
            self.env['account.asset.profile']._selection_method_period()

grindtildeath commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1075 @@
+        states={'draft': [('readonly', False)]}, default=0.3)
+    method_time = fields.Selection(
+        selection=lambda self: self.env['account.asset.profile'].
+        _selection_method_time(), string='Time Method',
```
same as above

grindtildeath commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1075 @@
+        inverse_name='asset_id',
+        string='Depreciation Lines', copy=False,
+        readonly=True, states={'draft': [('readonly', False)]})
+    type = fields.Selection(
```
Since we rename fields and models, can we use another name not to overwrite the builtin ?

grindtildeath commented:
@luc-demeyer I'm trying to do a code review but the diff is so huge I'm getting lost. :sweat_smile: 
So I'll try to correct what I can at the same time I'm reviewin, and I'll open a PR against your branch when I'm done :wink:

grindtildeath committed:
Merge branch '10-asset_management' of github.com:Noviat/account-financial-tools into 10.0_noviat_asset_management

grindtildeath committed:
Fix compute methods dependencies and small optimizations

grindtildeath reviewed (commented):
@luc-demeyer I opened https://github.com/Noviat/account-financial-tools/pull/3 with some small changes in it.
Would be great if you rebase this branch first onto OCA/10.0, so you can get only my own commits when merging.
Also, we need more reviews on this but I fear the diff scare people out.
Maybe can you leave only account_asset_management here and open a new PR with modules _import, _method_number_end and _xls ?

sanrav commented:
@luc-demeyer I will not be present at the OCA code sprint next week but I can dedicate some time to do some tests on Monday or Tuesday.

luc-demeyer committed:
fix typo

grindtildeath committed:
Rename demo file to test and move it into the right folder

luc-demeyer committed:
Merge pull request #3 from grindtildeath/10.0_noviat_asset_management

10.0 noviat asset management

luc-demeyer committed:
fix test data

luc-demeyer committed:
fix onchange

grindtildeath committed:
Fix tests on account_asset_management_xls and account_asset_management_method_number_end

luc-demeyer committed:
Merge pull request #4 from grindtildeath/10.0_noviat_asset_management_xls

Fix tests on account_asset_management_xls

grindtildeath commented:
Travis error is unrelated (fixed in https://github.com/OCA/account-financial-tools/pull/699)

luc-demeyer committed:
align with 11.0 port

luc-demeyer committed:
update views

arkostyuk commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1092 @@
+            for line in asset.depreciation_line_ids:
+                if line.move_id:
+                    asset.move_line_check = True
+                    continue
```
seems like `continue` is not needed here

grindtildeath replied:
Very good catch :+1:

arkostyuk commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1092 @@
+                    raise UserError(
+                        _("The Start Date must precede the Ending Date."))
+
+    @api.onchange('purchase_value', 'salvage_value', 'date_start', 'method')
```
lacks `depreciation_base`?

grindtildeath replied:
Not needed, since `self.depreciation_base` is calculated inside the onchange and the triggering fields are the good ones. :wink:

arkostyuk commented on account_asset_management/models/account_asset.py:
```diff
@@ -0,0 +1,1092 @@
+                    factor = float(duration) / cy_days
+                elif i == cnt - 1:  # last year
+                    duration = (
+                        fy_date_stop - datetime(year, 01, 01)).days + 1
```
not blocking, but `01` gives you an octal number in return - unless you're not planning to use months/days greater than 7

arkostyuk commented on account_asset_management/wizard/account_asset_compute.py:
```diff
@@ -0,0 +1,68 @@
+            }
+
+        domain = "[('id', 'in', [" + \
+            ','.join(map(str, created_move_ids)) + "])]"
```
why cast it in a string?

arkostyuk commented on account_asset_management/wizard/account_asset_remove.py:
```diff
@@ -1,188 +1,242 @@
+    @api.model
+    def _selection_posting_regime(self):
+        return[
+            ('residual_value', _('Residual Value')),
```
I doubt that you should translate those here

grindtildeath replied:
Just tested, and still needed :-1:

grindtildeath committed:
Do not use builtin type in function

luc-demeyer committed:
fix _compute_move_line_check

luc-demeyer committed:
datetime(year, 01,01) to datetime(year, 1, 1)

luc-demeyer committed:
simplify domain

grindtildeath commented:
@sanrav @arkostyuk Since you reviewed, can we have your approval ?

sanrav commented:
Tested on #698 
When you go on the menu "Asset hierarchy", ODOO opens the form view of a new asset. It should show the hierarchy view.
I don't see the menu to export the xls file "Financial assets report".

grindtildeath committed:
Fix KeyError when invl_id does not exist

grindtildeath commented:
@sanrav This PR is for version 10.0.
So asset hierarchy as well as the xls button are there in this version, and what concerns v11.0 should be stated on #698 and not here.
Hierarchical tree views are not supported in 11.0 so I did add a commit to drop it.
Regarding the XLS button, it won't be available in #698 neither, because the XLS module is migrated in #701, so you will see it there as soon as a runbot will be available.
Therefore, your approval would be much appreciated here, and a review on the following PRs as well. :wink:

luc-demeyer commented:
@sbidoul 
work finished on this PR and tested intensively technical as well as functional. 
Can you merge ?
Tx,
Luc

sbidoul commented:
@luc-demeyer thanks for this effort!

luc-demeyer commented:
merge please.
